### PR TITLE
Feature/ppca

### PR DESCRIPTION
**Pull request by austinTalbot7241993**
This implements the basic maximum likelihood PPCA model along with a base class for all future PPCA/factor analysis models.

### CODE REVIEW

**Review by poneill — APPROVED** [review]
I would strongly recommend:

- fully initializing all objects explicitly in their __inits__, rather than deferring parts of initialization to private methods that (afaict?) aren't called
- inheriting from abc.ABC whenever you have an @abstractmethod, otherwise the class will not behave as you're intending
- publicizing methods that are imported by another module

But I defer to your judgment to balance those concerns against velocity.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        super().__init__(n_components=n_components)
+
+    def __repr__(self):
+        out_str = "PPCA_analytic object\n"
+        out_str += "n_components=%d\n" % self.n_components
+        return out_str
+
+    def fit(self, X):
+        """
+        Fits a model given covariates X as well as option labels y in the
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:40]: Artifact from when I made every fit method include unused arguments like sklearn's style, will remove.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        Analytic PPCA solution as described by Bishop
+
+        Parameters
+        ----------
+        n_components : int,default=2
+            The latent dimensionality
+        """
+        super().__init__(n_components=n_components)
+
+    def __repr__(self):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:33]: This is mostly a surprise because Python’s defaults tend to be fairly useful. However, in this case, having a default for __repr__ which would act like:

return "%s(%r)" % (self.__class__, self.__dict__)
would have been too dangerous (for example, too easy to get into infinite recursion if objects reference each other). So Python cops out. Note that there is one default which is true: if __repr__ is defined, and __str__ is not, the object will behave as though __str__=__repr__.

This means, in simple terms: almost every object you implement should have a functional __repr__ that’s usable for understanding the object. Implementing __str__ is optional: do that if you need a “pretty print” functionality (for example, used by a report generator).

I'm really lazy and would prefer not to make two methods. But willing to be overruled.

File: python/python/bystro/supervised_ppca/_base.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+        if isinstance(log_prior, np.ndarray):
+            self.losses_prior[i] = log_prior
+        else:
+            self.losses_prior[i] = log_prior.numpy()
+        self.losses_posterior[i] = log_posterior.numpy()
+
+    @abstractmethod
+    def _create_prior(self):
+        """
+        Creates a prior on the parameters taking your trainable variable
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:505]: Shoot, good catch.

File: python/python/bystro/supervised_ppca/_base.py [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+        ----------
+        n_components : int,default=2
+            The latent dimensionality
+
+        Sets
+        ----
+        creationDate : datetime
+            The date/time that the object was created
+        """
+        self.n_components = int(n_components)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:68]: Gotten burned in the past where 1.0 returns an error.

File: python/python/bystro/supervised_ppca/_base.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+        None
+
+        Returns
+        -------
+        covariance : np.array-like(p,p)
+            The covariance matrix
+        """
+
+    @abstractmethod
+    def _save_variables(self, trainable_variables):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:106]: OK can change name. I try and avoid massive methods and if I don't separate out the initialize variables, initialize losses we are tracking, test inputs (will be in future code), and saving variables we end up with enormous fit methods that are harder to read. If you want, I can show you in the future what things look like if we don't do this.

File: python/python/bystro/supervised_ppca/_base.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+        srank = _get_stable_rank(covariance)
+        return srank
+
+    def transform(self, X):
+        """
+        This returns the latent variable estimates given X
+
+        Parameters
+        ----------
+        X : np array-like,(N_samples,\sum idxs)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:158]: What this means is that you pass in X as observed variables, not the shape of the fit variables. Example a.

suppose idxs = [1,1,0]. Then X should be a Nx2 matrix.

File: python/python/bystro/supervised_ppca/_base.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            The data to transform
+
+        Returns
+        -------
+        S : np.array-like,(N_samples,n_components)
+            The factor estimates
+        """
+        prec = self.get_precision()
+        coefs = np.dot(self.W_, prec)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:167]: Grew up old school before @ existed. The *very tiny* advantage of dot is that it's easier to control the order of evaluations of multiplications which can save time depending on matrix shapes. Do we want to make @ a standard?

File: python/python/bystro/supervised_ppca/_base.py [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+        Returns
+        -------
+        avg_score : float
+            Average log likelihood
+        """
+        if sherman_woodbury is False:
+            covariance = self.get_covariance()
+            avg_score = _conditional_score(covariance, X, idxs, weights=weights)
+        else:
+            raise NotImplementedError("Sherman woodbury not imlemented yet")
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:197]: I don't understand

File: python/python/bystro/supervised_ppca/_base.py [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+
+        log p(X[idx==1]|X[idx==0],covariance)
+
+        Parameters
+        ----------
+        X : np.array-like,(N,p)
+            The data
+
+        idxs: np.array-like,(p,)
+            The observation locations
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:212]: OK can change.

File: python/python/bystro/supervised_ppca/_base.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+        """
+        default_options = {
+            "n_iterations": 3000,
+            "learning_rate": 1e-2,
+            "gpu_memory": 1024,
+            "method": "Nadam",
+            "batch_size": 100,
+            "momentum": 0.9,
+        }
+        tops = {**default_options, **training_options}
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:453]: Wow that's nifty.

File: python/python/bystro/supervised_ppca/_base.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+        return tops
+
+    def _initialize_saved_losses(self):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:456]: It is not called in analytic ppca since that has a closed-form solution. I figured it wasn't a problem since it should never be called by the user.

File: python/python/bystro/supervised_ppca/_base.py [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+            The log likelihood
+
+        losses_prior : tf.Float
+            The log prior
+
+        losses_posterior : tf.Float
+            The log posterior
+        """
+        self.losses_likelihood[i] = log_likelihood.numpy()
+        if isinstance(log_prior, np.ndarray):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:496]: Oops, this got ported from tensorflow. And this is more because sometimes log_prior is a float and sometimes a torch.tensor

File: python/python/bystro/supervised_ppca/_base.py [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+        dictionary as input
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        log_prior : function
+            The function representing the negative log density of the prior
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:515]: stochastic gradient descent of negative log likelihood.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        Returns
+        -------
+        self : object
+            The model
+        """
+        N, self.p = X.shape
+        L, p = self.n_components, self.p
+
+        U, s, V = la.svd(X, full_matrices=False)
+        evals = s ** 2 / (N - 1)
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:57]: Will change.

File: python/python/bystro/supervised_ppca/_base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/_base.py:197]: sorry, that was unclear on my part.  I'm suggesting a change to the error message so that the line reads:

```
raise NotImplementedError("Sherman-Woodbury matrix inversion not implemented yet")
```

(or at least we should fix the typo imlemented -> implemented)

File: python/python/bystro/supervised_ppca/_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/_base.py:167]: I def agree that we sometimes have to be careful about how to order matrix multiplications in order to avoid storing large intermediate values.  For example, if we have four matrices A (N x N), B (N x N) and C (N x N) and D (N x m) with N >> m, then we'll want to be sure to compute the matrix product ABCD as (A (B (C D))) in order to avoid directly computing AB or BC.

With the matrix mult operator that looks like:

```python
 (A @ (B @ (C @ D)))
```

and with np.dot it looks like:
```python
 np.dot(A, np.dot(B, np.dot(C, D)))
```

So we get the same control, but the matrix mult operator is (imho) easier to both read and type, YMMV :)

File: python/python/bystro/supervised_ppca/_base.py [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+    gradient descent for inference. This reduces boilerplate code 
+    associated with some of the standard methods etc.
+
+Methods
+-------
+None
+"""
+from abc import abstractmethod
+import numpy as np
+from bystro.covariance._base_covariance import ( # type: ignore
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:35]: Stubs I believe. I haven't done typing in any of my files.

File: python/python/bystro/supervised_ppca/_base.py [hunk 16]
```diff
@@ -0,0 +1,516 @@
...
+    _score, # type: ignore
+    _score_samples, # type: ignore
+    _entropy, # type: ignore
+    _entropy_subset, # type: ignore
+    _mutual_information, # type: ignore
+) # type: ignore 
+from numpy import linalg as la
+from datetime import datetime as dt
+import pytz
+from bystro._template_sgd_np import _BaseSGDModel # type: ignore
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:50]: Yeah will get rid of underscore.

File: python/python/bystro/supervised_ppca/_base.py [hunk 17]
```diff
@@ -0,0 +1,516 @@
...
+
+
+class BaseGaussianFactorModel(_BaseSGDModel):
+    def __init__(self, n_components=2):
+        """
+        This is the base class of the model. Will never be called directly.
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:56]: Double inheritance?

File: python/python/bystro/supervised_ppca/_base.py [hunk 18]
```diff
@@ -0,0 +1,516 @@
...
+
+        Sets
+        ----
+        creationDate : datetime
+            The date/time that the object was created
+        """
+        self.n_components = int(n_components)
+        self.creationDate = dt.now(pytz.timezone("US/Pacific"))
+
+    @abstractmethod
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:71]: Added

File: python/python/bystro/supervised_ppca/_base.py [hunk 19]
```diff
@@ -0,0 +1,516 @@
...
+        n_components : int,default=2
+            The latent dimensionality
+
+        Sets
+        ----
+        creationDate : datetime
+            The date/time that the object was created
+        """
+        self.n_components = int(n_components)
+        self.creationDate = dt.now(pytz.timezone("US/Pacific"))
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:69]: Switched it

File: python/python/bystro/supervised_ppca/_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:106]: Will push an actual model that uses these options in a future pr before we make the final decision.

File: python/python/bystro/supervised_ppca/_base.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:158]: Added comment.

File: python/python/bystro/supervised_ppca/_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:167]: Will possibly change in the future if that's OK.

File: python/python/bystro/supervised_ppca/_base.py [hunk 20]
```diff
@@ -0,0 +1,516 @@
...
+        idxs: np.array-like,(sum(p),)
+            The observation locations
+
+        Returns
+        -------
+        scores : float
+            Average log likelihood
+        """
+        if sherman_woodbury is False:
+            self.get_covariance()
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:271]: Don't know why it got deleted.

File: python/python/bystro/supervised_ppca/_base.py [hunk 21]
```diff
@@ -0,0 +1,516 @@
...
+        Returns
+        -------
+        entropy : float
+            The differential entropy of the distribution
+        """
+        covariance = self.get_covariance()
+        entropy = _entropy(covariance)
+        return entropy
+
+    def entropy_subset(self, idxs):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:340]: Done.

File: python/python/bystro/supervised_ppca/_base.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:453]: Ah this is where the comment was, thought it was on teams for some reason.

File: python/python/bystro/supervised_ppca/_base.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:515]: Because we minimize negative log posterior = -log_prior + -log_likelihood.

File: python/python/bystro/supervised_ppca/_base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:505]: No, it will use self.prior_options.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:33]: I'm lazy and would rather just implement a single method.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:40]: Got rid of y.

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:57]: Done.

File: python/python/bystro/supervised_ppca/_base.py [hunk 22]
```diff
@@ -0,0 +1,516 @@
...
+        -------
+        mutual_information : float
+            The mutual information between the two variables
+        """
+        covariance = self.get_covariance()
+        mutual_information = _mutual_information(covariance, idxs1, idxs2)
+        return mutual_information
+
+
+class BaseGDModel(BaseGaussianFactorModel):
```
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:382]: Done.

File: python/python/bystro/supervised_ppca/_base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/_base.py:35]: I remember it wasa complaining about something else.

File: python/python/bystro/supervised_ppca/_base.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/_base.py:543]: ok, my recommendation would be to document that expectation in the docstring then

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (poneill) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:33]: ok, for consistency with sklearn I would recommend just returning `f'PPCAAnalytic(n_components={self.n_components})'` then.  Sklearn objects return a string formatted in that style for both str and repr

**Review by poneill — APPROVED** [review]

File: python/python/bystro/supervised_ppca/ppca_analytic_np.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (austinTalbot7241993) [reply python/python/bystro/supervised_ppca/ppca_analytic_np.py:33]: OK sounds good.
